### PR TITLE
Fix broadcast InferCorrectLayout

### PR DESCRIPTION
**Pull request by lazycal**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

I'm reviewing this function after PR #10118, and found several buggy places. For instance, 
- it does not respect the original input tensor's layout but always reset it to the other input's layout (see `test_broadcast_respect_input_layouts`);
- it assumes one can always only transform one layout to adapt to the other. However, this is not always possible, see `test_broadcast_non_adaptable`.
- `AdjustSubordinateFactors` should also be called in the else branch https://github.com/apache/tvm/blob/6a274af9cb4e7baf6d9dd06f0fb38bea5ac3154a/src/relay/transforms/infer_layout_utils.h#L222-L234. (see `test_alter_layout_nonscalar_broadcast`)

This PR rewrites `BinaryBroadcastLayoutHelper` a bit, and I tried to make it as backward compatible as possible.

## Misc
BTW I seem to spot another bug when I was working on this PR. So I sometimes I need to create a scalar layout, and I used `Layout("")`, but it returned an undefined layout. From the doc it should return a scalar layout right? I checked the constructor, and it seems in this if statement: https://github.com/apache/tvm/blob/6a274af9cb4e7baf6d9dd06f0fb38bea5ac3154a/src/tir/ir/data_layout.cc#L99, it directly returns without setting the `data_` field as in normal path: https://github.com/apache/tvm/blob/6a274af9cb4e7baf6d9dd06f0fb38bea5ac3154a/src/tir/ir/data_layout.cc#L144. I didn't fix this issue but work it around somehow, but it does seem like a bug to me :-).

### CODE REVIEW

**Comment by lazycal** [comment]
@masahi

File: src/relay/transforms/infer_layout_utils.cc [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        auto new_shape_val = src_layout.FactorOf(dual_axis);
+        new_layout += std::to_string(new_shape_val);
+      }
+    }
+    new_layout += LayoutAxis::Get(axis).name();
+  }
+  return new_layout != "" ? Layout(new_layout)
+                          : Layout("H").SubLayout(0, 0);  // hack to create a scalar layout
+}
+bool Isomorphic(const Layout& lhs, const Layout& rhs) {
```
Comment (masahi) [inline src/relay/transforms/infer_layout_utils.cc:84]: new line

File: src/relay/transforms/infer_layout_utils.cc [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+    it = map_back.find(name_rhs);
+    if (it == map_back.end())
+      map_back[name_rhs] = name_lhs;
+    else if (it->second != name_lhs)
+      return false;
+    if (!lhs_axis.IsPrimal() && lhs.FactorOf(lhs_axis) != rhs.FactorOf(rhs_axis)) return false;
+  }
+  return true;
+}
```
Comment (masahi) [inline src/relay/transforms/infer_layout_utils.cc:112]: new line

File: src/relay/transforms/infer_layout_utils.cc [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    if (!orthorgonal || !Isomorphic(large_sublayout, small))
+      return Layout::Undef();  // For now this case is not supported.
+  }
+
+  // `old` is compatible. Now learn the axis name mapping between `old` and `ref_old`
+  if (old.ndim() == 0) return old;  // an optmization for scalar: no-op
+  int mapping[26];
+  bool used[26];
+  memset(mapping, -1, sizeof mapping);
+  memset(used, 0, sizeof used);
```
Comment (masahi) [inline src/relay/transforms/infer_layout_utils.cc:142]: use std::vector

File: src/relay/transforms/infer_layout_utils.cc [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  }
+  for (int i = ref_old->axes.size() - 1; i >= 0; --i) {
+    char name_ref = LayoutAxis::Get(ref_old->axes[i]).ToPrimal().name()[0];
+    int name = mapping[name_ref - 'A'];
+    if (name == -1) {
+      mapping[name_ref - 'A'] = find_unused(name_ref) - 'A';
+      used[mapping[name_ref - 'A']] = true;
+    }
+  }
+
```
Comment (masahi) [inline src/relay/transforms/infer_layout_utils.cc:164]: Add new lines above between for loop

**Comment by lazycal** [comment]
@masahi The code is changed according to your suggestions.

**Review by masahi — APPROVED** [review]
I didn't actually try to understand the code, but why not, ship it

**Pull request merged**